src/app/util/attribute-storage-null-handling.h

```diff
@@ -55,14 +55,49 @@ struct NumericAttributeTraits
     // interesting things here.
     static constexpr WorkingType StorageToWorking(StorageType storageValue) { return storageValue; }
 
+private:
+    // We need to make sure we never look like we are assigning NaN to an
+    // integer, even in a not-reached branch.  Without "if constexpr", the best
+    // we can do is these functions using enable_if.
+    template <typename U = T, typename std::enable_if_t<std::is_floating_point<U>::value, int> = 0>
+    static constexpr StorageType GetNullValue()
+    {
+        return std::numeric_limits<T>::quiet_NaN();
+    }
+
+    template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value, int> = 0>
+    static constexpr StorageType GetNullValue()
+    {
+        return std::is_signed<T>::value ? std::numeric_limits<T>::min() : std::numeric_limits<T>::max();
+    }
+
+    template <typename U = T, typename std::enable_if_t<std::is_enum<U>::value, int> = 0>
+    static constexpr StorageType GetNullValue()
+    {
+        return GetNullValue<std::underlying_type_t<T>>();
+    }
+
+public:
     // The value reserved in the value space of StorageType to represent null,
     // for cases when we have a nullable value.  This value must match the value
     // excluded from the valid value range in the spec, so that we don't confuse
     // valid values with null.
-    static constexpr StorageType kNullValue =
-        std::is_signed<T>::value ? std::numeric_limits<T>::min() : std::numeric_limits<T>::max();
+    static constexpr StorageType kNullValue = NumericAttributeTraits::GetNullValue();
 
-    static constexpr bool IsNullValue(StorageType value) { return value == kNullValue; }
+    template <typename U = T, typename std::enable_if_t<!std::is_floating_point<U>::value, int> = 0>
+    static constexpr bool IsNullValue(StorageType value)
+    {
+        return value == kNullValue;
+    }
+
+    template <typename U = T, typename std::enable_if_t<std::is_floating_point<U>::value, int> = 0>
+    static constexpr bool IsNullValue(StorageType value)
+    {
+        // Trying to include math.h (to use isnan()) fails on EFR32, both when
+        // included as "cmath" and when included as "math.h".  For lack of
+        // isnan(), just fall back on the NaN != NaN thing.
+        return value != value;
```

Comment:
I believe that it would be entirely legal for a platform to hit a math trap on this, even if std::numeric_limits<T>::quiet_NaN() exists.  Do we care?  I don't know that such a thing would necessarily be caught in unit tests.

To be fair, NaN is obviously a more appropriate value for NULL than numeric_limits<float>::min().

Reply:
I would be very very happy to do something better here.  I'm just not sure what, if `isnan()` is not usable.  Maybe the right answer is to add a TODO and file a followup to use isnan here once we sort out whatever the efr32 bugginess in math.h is?

And yes, NaN is what the spec says to use for null (not in so many words).

Reply:
Also, I think we need std::numeric_limits<T>::is_iec559 if we want to know that float is really an IEEE-754 single and double is really an IEEE-754 double.  Also, quiet_NaN is platform specific.  Is this something that can come across the wire?

Reply:
Sorry, cross-posting.

> And yes, NaN is what the spec says to use for null (not in so many words).

OK, I didn't realize that.  NaN is unambiguous assuming IEEE-754 (however, quiet_NaN isn't necessarily).

TODO seems appropriate.  I would only point out that:

* IEEE-754 is not guaranteed unless checking e.g. is_iec559
* assuming IEEE754, width isn't guaranteed unless checking e.g. __DBL_MANT_DIG__ or is_iec559
* math traps on comparison are possible unless you know the platform and set the config bits yourself
* there are lots of buggy / non-compliant corners for embedded platforms when it comes to infinity / NaN

Given all of this, it seems like it would be a good idea to verify that the float and double types are what we expect.  As for evaluation of NaN, in the past I have resorted to bit-inspection for platforms that don't properly support this.  To this, I think we should use isnan where it's available and only conditionally compile an alternative (the != trick or possibly looking at the bits directly) when needed.

Reply:
Filed https://github.com/project-chip/connectedhomeip/issues/12453 to track this.